torch/distributed/distributed_c10d.py

```diff
@@ -1297,6 +1298,65 @@ def get_node_local_rank(fallback_rank: Optional[int] = None) -> int:
     )
 
 
+def _get_backends_for_timeout_set(group: ProcessGroup) -> Set[c10d_Backend]:
+    """
+    Return the all possible backends for a given process group when we
+    need to set the timeout for a ProcessGroup.
+
+    Args:
+        group (ProcessGroup): The process group to be set timeout.
+
+    Returns:
+        set[c10d_Backend]: A set of backends which we can set timeout for.
+    """
+    assert isinstance(group, ProcessGroup)
+    devices = group._device_types
+    backends = set()
+    if torch.device("cpu") in devices and is_gloo_available():
+        backend = group._get_backend(torch.device("cpu"))
+        if isinstance(backend, ProcessGroupGloo):
+            backends.add(backend)
+    if torch.device("cuda") in devices:
+        backend = group._get_backend(torch.device("cuda"))
+        if is_nccl_available() and isinstance(backend, ProcessGroupNCCL):
+            backends.add(backend)  # type: ignore[arg-type]
+        elif is_gloo_available() and isinstance(backend, ProcessGroupGloo):
+            backends.add(backend)  # type: ignore[arg-type]
+    if len(backends) == 0:
+        warnings.warn("Set timeout is now only supported for either nccl or gloo.")
+    return backends  # type: ignore[return-value]
+
+
+def _extend_timeout_all_pgs(timeout: timedelta) -> None:
+    """
+    This API extends the timeout for all PGs locally on one rank.
+    If user sets a very large negative timeout, we will set the timeout
+    to be the negative value.
```

Comment:
Hmm, we should never allow a negative timeout?

Reply:
I'm not sure it's a good idea to let people subtract from the timeout.